### PR TITLE
Config refactoring

### PR DESCRIPTION
**Pull request by Trishia**
I just pushed config refactoring of coap-gateway and authorization on config-refactoring branch but the others are not yet [Latest commit : plgd-dev/cloud@9f6d6ea]. If you have time to review, inform me your opinion regarding to yaml, code cleansing and so on.

### CODE REVIEW

File: coap-gateway/coap-gw.yaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
+log:
+  debug: true
+apis:
+  coap:
+    address: "0.0.0.0:5684"
+    external-port: 5684
+    fqdn: "localhost"
```
Comment (jkralik) [inline coap-gateway/coap-gw.yaml]: these 2 settings we want to replace by `externalAddress` -> it is in format `fqdn:external-port`

File: coap-gateway/coap-gw.yaml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    request-timeout: 10s
```
Comment (jkralik) [inline coap-gateway/coap-gw.yaml]: request-timeout -> timeout

File: coap-gateway/coap-gw.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    heartbeat: 4s
+    max-message-size: 1024  # default:262144
+    keepalive-enabled: true
+    keepalive-timeout: 20s
+    blockwise-transfer-enabled: true
+    blockwise-transfer-block-size: 1024
```
Comment (jkralik) [inline coap-gateway/coap-gw.yaml]: move to capabilities section

File: coap-gateway/coap-gw.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+clients:
+  oauth-provider:
+    oauth:
+      client-id: ""
+      client-secret: ""
+      scopes: []
+      audience: ""
+      endpoint:
+        token-url : ""
+      request-timeout : 10s
```
Comment (jkralik) [inline coap-gateway/coap-gw.yaml]: move to `coap.authorization` like:
```
apis.coap.authorization:
      authority: ""
      audience: ""
```

File: coap-gateway/coap-gw.yaml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    tls:
+      enabled: true
+      ca-file: "/data/certs/root_ca.crt"
+      dir-path: "/data/certs"
+      cert-file: "http.crt"
+      key-file: "http.key"
+      use-system-ca-pool: true
+  authorization:
```
Comment (jkralik) [inline coap-gateway/coap-gw.yaml]: authorization  -> authorization-server

File: coap-gateway/coap-gw.yaml [hunk 6]
```diff
@@ -0,0 +1,68 @@
+log:
```
Comment (jkralik) [inline coap-gateway/coap-gw.yaml]: expected coap configuration(@ondrejtomcik):
```
log:
  debug: false
clients:
  resourceAggregate:
    address: ""
    tls:
      enabled: true
      caFile: ""
      certFile: ""
      keyFile: ""
      useSystemCAPool: false
  resourceDirectory:
    address: ""
    tls:
      enabled: true
      caFile: ""
      certFile: ""
      keyFile: ""
      useSystemCAPool: false
  authorizationServer:
    address: ""
    tls:
      enabled: true
      caFile: ""
      certFile: ""
      keyFile: ""
      useSystemCAPool: false
apis:
  coap:
    address: ""
    externalAddress: ""
    timeout: 0s 
    goroutineSocketHeartbeat: 4s
    capabilities:
      maxMessageSize: 1024
      keepalive:
        enabled: true
        timeout: 5s
        maxRetries: 6
      blockwiseTransfer:
        enabled: true
        blockSize: 1024
    tls:
      enabled: true
      caFile: ""
      certFile: ""
      keyFile: ""
      clientCertificateRequired: true
    authorization:
      authority: ""
      audience: ""
```

File: authorization/authorization.yaml [hunk 7]
```diff
@@ -0,0 +1,76 @@
+log:
```
Comment (jkralik) [inline authorization/authorization.yaml:1]: expected authorization configuration:
```
log:
  debug: false
oAuthClients:
  device:
    provider: ""
    providerName: ""
    clientID: ""
    clientSecret: ""
    scopes: []
    authorizationURL: ""
    tokenURL: ""
    redirectURL: ""
  client:
    clientID: ""
    clientSecret: ""
    scopes: []
    authorizationURL: ""
    tokenURL: ""
    redirectURL: ""
database:
  mongoDB:
    uri: ""
    database: ""
    tls:
      caFile: ""
      certFile: ""
      keyFile: ""
      clientCertificateRequired: true
apis:
  http:
    address:
    tls:
      caPool: ""
      tlsCert: ""
      tlsKey: ""
      clientCertificateRequired: true
    authorization:
      authority: ""
      audience: ""
  grpc:
    address:
    tls:
      caPool: ""
      tlsCert: ""
      tlsKey: ""
      clientCertificateRequired: true
    authorization:
      authority: ""
      audience: ""
```

File: coap-gateway/coap-gw.yaml (hunk 1 above)
Comment (Trishia) [reply coap-gateway/coap-gw.yaml]: You means you want to merge two settings "external-port" and "fqdn" with one setting "externalAddress" ? Sure to remove fqdn ?

File: coap-gateway/coap-gw.yaml (hunk 4 above)
Comment (Trishia) [reply coap-gateway/coap-gw.yaml]: Give me more detail. Your are using below codes for coap.authorization. In this case OAuth config is needed but two authority and audience. 
```
// ToClientCrendtials converts to clientcredentials.Config
func (c Config) ToClientCrendtials() clientcredentials.Config {
	v := make(url.Values)
	if c.Audience != "" {
		v.Set("audience", c.Audience)
	}

	return clientcredentials.Config{
		ClientID:       c.ClientID,
		ClientSecret:   c.ClientSecret,
		Scopes:         c.Scopes,
		TokenURL:       c.Endpoint.TokenURL,
		EndpointParams: v,
	}
}
```

File: coap-gateway/coap-gw.yaml (hunk 1 above)
Comment (jkralik) [reply coap-gateway/coap-gw.yaml]: yes

File: coap-gateway/coap-gw.yaml (hunk 4 above)
Comment (jkralik) [reply coap-gateway/coap-gw.yaml]: You are right it is for verify token. So leave it.

File: coap-gateway/coap-gw.yaml (hunk 3 above)
Comment (Trishia) [reply coap-gateway/coap-gw.yaml]: Ok, I got it.  I will move that to capabilities. 
By the way, keepalive and blockwise should be also sub structure as you mentioned below comment of expected coap configuration ? I think capabilities structure is enough to include all that.

**Comment by Trishia** [comment]
@jkralik @ondrejtomcik 
Now, authorization, certificate-authority, resource-aggregate, resource-directory, coap-gateway, grpc-gateway, http-gateway except c2c-gateway and c2c-connector were tested as a container in bridge network. Review plz.

**Comment by ondrejtomcik** [comment]
TODO @jkralik : drop refImpl and extend config interface to SetDefaults()

**Comment by Trishia** [comment]
@jkralik @ondrejtomcik Now, I pushed new commit [https://github.com/plgd-dev/cloud/pull/148/commits/d2404e87880ad61e05c28af0aafbbed044d3d53e] that cloud2cloud-connector, cloud2cloud-gateway were tested in bridge network. Review plz.

**Comment by Trishia** [comment]
@jkralik @ondrejtomcik  Do I have to refactor portal-webapi's config? I don't know the use of portal-webapi for v2.

**Comment by ondrejtomcik** [comment]
> @jkralik @ondrejtomcik Do I have to refactor portal-webapi's config? I don't know the use of portal-webapi for v2.

Hello @Trishia. No, portal-webpi is obsolete and will be completely removed. New UI is in progress, you can already check it on try.plgd.cloud .

**Comment by Trishia** [comment]
This pull requesting of config-refactoring branch will be closed because it is no need any more.

**Pull request closed without merging**